ethcore/src/engines/clique/block_state.rs

```diff
@@ -0,0 +1,207 @@
+// Copyright 2015-2019 Parity Technologies (UK) Ltd.
+// This file is part of Parity Ethereum.
+
+// Parity Ethereum is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity Ethereum is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity Ethereum.  If not, see <http://www.gnu.org/licenses/>.
+
+use std::collections::{HashMap, VecDeque};
+use std::time::{Duration, SystemTime};
+use std::time::UNIX_EPOCH;
+
+use ethereum_types::Address;
+use rand::Rng;
+
+use engines::clique::{VoteType, DIFF_INTURN, DIFF_NOTURN, NONCE_AUTH_VOTE, NONCE_DROP_VOTE, NULL_AUTHOR, SIGNING_DELAY_NOTURN_MS};
+use engines::clique::util::{extract_signers, recover_creator};
+use error::Error;
+use types::header::Header;
+
+#[derive(Clone, Debug)]
+pub struct CliqueBlockState {
+	pub votes: HashMap<(Address, Address), VoteType>, // k: (Voter, beneficiary), VoteType)
+	pub votes_history: Vec<(u64, Address, VoteType, Address)>, // blockNumber, Voter, VoteType, beneficiary
+	pub signers: Vec<Address>,
+	pub recent_signers: VecDeque<Address>,
+	pub next_timestamp_inturn: Option<SystemTime>, // inturn signing should wait until this time
+	pub next_timestamp_noturn: Option<SystemTime>, // noturn signing should wait until this time
+}
+
+impl CliqueBlockState {
+	/// Create new state with given information, this is used creating new state from Checkpoint block.
+	pub fn new(_author: Address, signers_sorted: Vec<Address>) -> Self {
```

Comment:
is `author` not needed for the constructor then I suggest to remove it instead

Reply:
Also I suggest to make `derive(Default)` for CliqueBlockState 

The you can impl `new` as:

```rust
pub fn new(signers_sorted) -> Self
   Self {
      signers: signers_sorted,
      ..Default::default()
   }
}
```